### PR TITLE
Refactor repository scaffolding

### LINKED ISSUE #104: Consider renaming repo or splitting `warehouse` sub-repo
https://github.com/E3SM-Project/esgfpub/pull/97#issuecomment-1050262215

> Another alternative is to separate the warehouse directory from the esgfpub repo in a new e3sm_warehouse repo. I recall Sterling saying like "it probably should be in its own repo but I added it here" or something to that effect.

https://github.com/E3SM-Project/esgfpub/pull/97#issuecomment-1050395831

> I am also leaning towards starting a new repo called e3sm_warehouse. I can actually make a copy of this repo as the starting point of the e3sm_warehouse repo, and just delete the /esgfpub directory there. Then we can archive the esgfpub repo if it is no longer being used.
> 
> As Jill mentioned, I would make a branch in this repo and continue working as needed (please no more work on master!). There are ways to move branches between repositories: https://stackoverflow.com/questions/34384642/how-to-copy-a-branch-from-one-github-repository-to-another

### LINKED ISSUE #110: Fix dependency conflicts involving `autocurator`
## Overview
- `autocurator=0.1` is only available for Linux and does not support `python>3.8`
- `autocurator=0.1` requires `libnetcdf >=4.7.4,<4.7.5.0a0`, which is not compatible with `nco>=5`
- As a result of these conflicts, `esgf-forge` dependencies could not be included in `prod.yml`

## The Problem
This means we have to support two separate conda-envs, one for data processing and one for data publishing (which requires `esgf-forge` dependencies such as `autocurator`). 

It also blocks up from being able to release this repository as an Anaconda package since the dependencies aren't compatible with one another (refer to #99).

## Possible Solutions
1. Find an alternative to `autocurator` since it is not being actively maintained
2. Work with `autocurator` developers to update dependencies 
  - Per Sasha:
    >   It would be a bit of work on my end and I can’t promise it by a particular date, but another option for the publisher is to replace autocurator with an alternative (eg. xarray to scan the metadata for each file).  Replacement would be unfortunate because that means reinventing the wheel for the key functionality to reconcile variables and metadata across a series of files: exactly what’s needed to define a dataset for publication.

### PR DESCRIPTION
**Pull request by tomvothecoder**
## Overview
This PR updates the scaffolding of the repository to follow Python conventions. It also removes the `esgfpub` directory since it is no longer being used.

## Summary of changes
- Move `/warehouse` directory one level down to the root of the repo.
  - This directory was previously `/warehouse/warehouse`.
  - Python module imports were safely updated using PyCharm.
- Rename `/warehouse` directory  to `/datasm`.
  - Python module imports were safely updated using PyCharm.
  - Update some `warehouse` string refs to `datasm`, except paths, command arguments, and classes.
- Remove `/esgfpub` directory and related refs.
  - Closes #104 

DevOps
- Move `/datasm/setup.py` to the root of repo and update contents.
  - Now we run `pip install . ` from the root of the repo, instead of `cd warehouse` first.
- Rename `dev.yml` to `proc.yml`. 
  - Update references for "development" to "processing"
  - It is more accurate to call this environment the "processing" env because the "publishing" env is also used for development.
- Remove `prod.yml` since it is not working yet. 
   - Need to resolve #110 and #99 first

Documentation
- Update `README.md` and `docs/getting_started.md`
- Move `NewCMIPVariableSetup.md` and `warehouse_processing.md` to `/docs`
- Update `.gitignore` with standard list from https://github.com/github/gitignore/blob/main/Python.gitignore

## What to review in this PR
I already performed a self-review of the items below, but more eyes are always. Most of the changes in the diff are related to deleting the `/esgfpub` directory, so you can ignore those

- [x] Make sure documentation updates are free of error (mostly path updates)
- [x] Local `datasm` build still builds and installs in the processing and publishing envs
- [x] Local `datasm` build runs the same as before using `datasm` command, instead of `warehouse`
  - Example: `datasm --help`, `datasm validate --help`

## How to test this PR

Processing operations -- `extract`, `validate`, `postprocess`
1. Install and activate `datasm_proc` env
   - `conda env install -f conda-env/proc.yml`
   - `conda activate datasm_proc`
2. Install local build into `datasm_proc` env
   - Make sure you are in the root of `esgfpub`
   - `pip install .`
3. Run `datasm` processing operations to make sure things are good to go
   - For example, confirm there are no `ModuleNotFound` errors

Publishing operations -- `process`
1. Install and activate `datasm_pub` env
   - `conda env install -f conda-env/pub.yml`
   - `conda activate datasm_pub`
2. Install local build into `datasm_pub` env
   - Make sure you are in the root of `esgfpub`
   - `pip install .`
3. Run `datasm` publishing operation to make sure things are good to go
   - For example, confirm there are no `ModuleNotFound` errors

## Next Actions
- [x] Rename the repository to `datasm`

### CODE REVIEW

**Comment by tomvothecoder** [comment]
I just realized there are merge conflicts with the latest commits, so this actually isn't ready for review yet.

File: e3sm_warehouse/e3sm_warehouse.py
Comment (tomvothecoder) [inline e3sm_warehouse/e3sm_warehouse.py:657]: I might need to revert these specific argument name changes since they probably don't make sense

File: e3sm_warehouse/resources/warehouse_config.yaml
Comment (tomvothecoder) [inline e3sm_warehouse/resources/warehouse_config.yaml:1]: Should we rename this directory to align with the repo and package?

File: e3sm_warehouse/e3sm_warehouse.py
Comment (tomvothecoder) [inline e3sm_warehouse/e3sm_warehouse.py:25]: Example module import updates

File: conda-env/ci.yml
Comment (tomvothecoder) [inline conda-env/ci.yml:3]: Added `e3sm_` suffix to conda env yml file names and prefixs

**Review by tomvothecoder — COMMENTED** [review]
A few comments

File: e3sm_warehouse/e3sm_warehouse.py
Comment (TonyB9000) [reply e3sm_warehouse/e3sm_warehouse.py:25]: This is another example of term-overloading.  We still refer to the user_pub filesystem location, to which archives datasets are extracted, as the "warehouse" (fully, /p/user_pub/e3sm/warehouse), and the "-w" of the command line allows a user to override that location.

File: e3sm_warehouse/e3sm_warehouse.py
Comment (TonyB9000) [reply e3sm_warehouse/e3sm_warehouse.py:657]: This is another example of term-overloading. We still refer to the user_pub filesystem location, to which archives datasets are extracted, as the "warehouse" (fully, /p/user_pub/e3sm/warehouse), and the "-w" of the command line allows a user to override that location.

File: e3sm_warehouse/resources/warehouse_config.yaml
Comment (TonyB9000) [reply e3sm_warehouse/resources/warehouse_config.yaml:1]: No, I'd want to avoid that.  It already has e3sm/warehouse in the path, and e3sm/e3sm_warehouse would be redundant.  Also, MANY scripts and config files rely upon it - that would be a lot of editing and testing.

File: e3sm_warehouse/e3sm_warehouse.py
Comment (TonyB9000) [reply e3sm_warehouse/e3sm_warehouse.py:25]: This is another matter. Granted, we have no other "warehouse" project, but changing to "e3sm_warehouse" is a possibility.  Major hassle, but if it makes something cleaner, it is do-able.

File: conda-env/ci.yml
Comment (TonyB9000) [reply conda-env/ci.yml:3]: I'm not sure of the impact here.  It would not hurt, I suspect.

File: e3sm_warehouse/e3sm_warehouse.py
Comment (TonyB9000) [reply e3sm_warehouse/e3sm_warehouse.py:657]: Note that there is a corresponding argument "-p", indicating the "publication" location of datasets that are published.  Often referred to as the "pub_dirs", that default is "/p/user_pub/work" an historical choice that I think is terrible, but necessary legacy access.

**Review by chengzhuzhang — APPROVED** [review]
Thanks for the PR. 
At some point, if Tony can have a test to run warehouse with these changes, I'd like to try as well

**Comment by TonyB9000** [comment]
We do have "datasets" we can "warehouse_extract", "warehouse_validate", "warehouse_generate", etc, that are small and could be good for high-level testing.  But someone will need to walk me through the steps I need to take to make this transition.  This includes any potential re-cloning of repos, any re-generation of conda environments, and then I will need to know what paths have changed (e.g. changing "warehouse" to "e3sm_warehouse") so I can search for them.  A lot of "quickie-hardcoded" stuff in scripts will need to be turned into ENV variables, so defined in one location.

File: e3sm_warehouse/resources/warehouse_config.yaml
Comment (tomvothecoder) [reply e3sm_warehouse/resources/warehouse_config.yaml:1]: Got it, that's what I was thinking too. I will revert my "find and replace" string changes that affected these paths.

File: conda-env/ci.yml
Comment (tomvothecoder) [reply conda-env/ci.yml:3]: This comment was just to make reviewers aware

File: e3sm_warehouse/e3sm_warehouse.py
Comment (tomvothecoder) [reply e3sm_warehouse/e3sm_warehouse.py:657]: Got it, I will revert these types of changes to `warehouse`. `e3sm_warehouse` will refer strictly to the Python package, not any paths.

File: datasm/dataset.py
Comment (tomvothecoder) [inline datasm/dataset.py:273]:
```suggestion
        # we assume that the warehouse directory contains only directories named "v0.#" or "v#"
```

File: datasm/warehouse.py
Comment (tomvothecoder) [inline datasm/warehouse.py:649]:
```suggestion
        p = parser.add_parser(name=NAME, help="Automated warehouse processing")
```

File: docs/getting_started.md
Comment (tomvothecoder) [inline docs/getting_started.md:126]:
```suggestion
  -w WAREHOUSE_PATH, --warehouse-path WAREHOUSE_PATH
```

File: docs/getting_started.md
Comment (tomvothecoder) [inline docs/getting_started.md:127]:
```suggestion
                        The root path for pre-publication dataset staging, default=/p/user_pub/e3sm/warehouse/
```

File: setup.py
Comment (tomvothecoder) [inline setup.py:60]:
```suggestion
    # TODO: Update this URL once the repo is renamed.
    url="https://github.com/E3SM-Project/esgfpub",
```

**Review by TonyB9000 — APPROVED** [review]
This looks good to me.  Thanks Tom.

**Comment by tomvothecoder** [comment]
> This looks good to me. Thanks Tom.

Thanks Tony. I pushed some new changes just now, which include renaming the `dev.yml` to `proc.yml`. Please review the latest PR description.

If you have time, it would be nice for you to test the new `datasm` command which supersedes `warehouse`. All of the subcommands should be the same. I provided instructions in the PR's description.

I would also like to merge this PR when it is least disruptive to your workflow, so that is dependent on your priority list.

**Comment by TonyB9000** [comment]
Hi Tom,  Did you want "proc" or "prod"?

I have a SLEW of updates to commit and merge to master.  How should I coordinate that with your repo-refactoring?

(I expect I will wipe/rename the old "esgfpub", and git-clone everything fresh.  THEN worry about external scripts vis-a-vis naming.

**Comment by tomvothecoder** [comment]
> Hi Tom, Did you want "proc" or "prod"?
>

@TonyB9000, I removed the `prod.yml` env for now to avoid confusion because a versioned release of this package doesn't exist yet. I also renamed `dev.yml` to `proc.yml`, which is the environment intended for "processing" operations (`extract, `validate,` postprocess`). 

I would like for you to test the local build (`pip install .`) using both the `proc.yml` and `pub.yml` and their related operations/commands.

> > I have a SLEW of updates to commit and merge to master. How should I coordinate that with your repo-refactoring?
> 
> (I expect I will wipe/rename the old "esgfpub", and git-clone everything fresh. THEN worry about external scripts vis-a-vis naming.)
> 

I think we should prioritize merging your changes to `master` first, then rebase this branch on top of `master`. I will update the affected files so that the imports reflect the new `datasm` name.

**Comment by TonyB9000** [comment]
OK, I will do a push and PR, so we can do this cut-over rather than drag it out.  My changes should incur no conflicts (none serious) and most recent are simply enhanced logging.

**Comment by tomvothecoder** [comment]
Hey @TonyB9000, this PR is ready for review. Please read through the description for the changelog, what to review, and how to test.

**Comment by TonyB9000** [comment]
Whole Lotta regression testing for name-changes in my "ancillary" codes and scripts - but to be expected...

**Comment by TonyB9000** [comment]
(I'm especially interested in the "How to Test" part.)

**Comment by tomvothecoder** [comment]
> How do I find the "changelog, what to review, and how to test"? (and I've much other stuff to test as well.)

The PR's description

**Comment by TonyB9000** [comment]
Excellent!  (You mean I have to read?  ;) ).  OK, here I go.

**Comment by TonyB9000** [comment]
The testing explains "Install and activate datasm_proc env".  But my (esgfpub) conda-env still has "dev.yml", etc.

Should I "git clone https://github.com/E3SM-Project/datasm" first?  Where is conda-env bundled?

**Comment by tomvothecoder** [comment]
> The testing explains "Install and activate datasm_proc env". But my (esgfpub) conda-env still has "dev.yml", etc.
and 
> Where is conda-env bundled?

You need to checkout the branch `devops/104-deprecate-esgfpub` to access the new conda env yml files.

> Should I "git clone https://github.com/E3SM-Project/datasm" first? 

The repo is going to be renamed to `datasm` after this PR is merged.

**Comment by TonyB9000** [comment]
OK, thanks.  Should "conda env install -f conda-env/proc.yml" be "conda env create"?

**Comment by TonyB9000** [comment]
Testing: So far so good!  At least, datasm -h, datasm validate -h, and datasm postprocess -h all produce the expected help-text.  Tomorrow I'll conduct some heavy testing.  I recall the "module load" errors were not triggered by running "help", but when loading the datasm subordinate modules with datasets.

**Comment by tomvothecoder** [comment]
Awesome, thanks Tony!

**Comment by tomvothecoder** [comment]
Hey @TonyB9000, how is testing going for this PR?

**Comment by TonyB9000** [comment]
Hey @tomvothecoder , all is fine with postprocessing (and I suspect anything invoked under "datasm <subprocess>").  I cannot test publication until I am ready to publish, and I have not yet tested many ancillary applications (consolidated status reporting) for naming or path issues - but they should be trivial to find and fix when the time comes.  I have a HUGE backlog of datasets to generate (and then, to publish) so my focus lies there,  By next Monday, I believe ALL of the "Amon" sets I am generating will have completed, and I will test "datasm publish" (and conda env "dsm_pub") at that time.  I can pause on Monday(ish) if you need to interrupt my job sequence.  (I *could* search all of my scripts for the term "warehouse" and see if any references to that module name remain).

**Comment by TonyB9000** [comment]
I have been "testing" (operating) dsm_proc and dsm_pub in the course of getting publication work done.  There is no way to thoroughly test, as there are too many one-off cases of things that need doing.  But I am satisfied that the name-changes and related restructuring have not broken anything serious.  In time, glitches will be uncovered and addressed.

Right now (in the middle of an ongoing set of post-process jobs) my only worry is that I should not "pip install" datasm while the plane in in flight.  In particular, it is calling e3sm_to_cmip routinely.  The current batch of jobs will finish before the Monday.

Perhaps I should (then) hold off starting new jobs until we can eliminate "devops/104-deprecate-esgfpub"?

**Comment by tomvothecoder** [comment]
Thanks @TonyB9000. Yes, let's hold off on further jobs until this PR is merged.

We should wait until the current batch of jobs are done Monday before we merge this PR.

**Comment by tomvothecoder** [comment]
@TonyB9000 I will go ahead and merge this PR. 

We will address any issues that pop up as we use `datasm`, but hopefully there are none since most changes are directory structure and name changes.

**Comment by tomvothecoder** [comment]
@TonyB9000 and @chengzhuzhang the repository has been renamed to "datasm". 

Your current local cloned repos with the previous "esgfpub" directory name will still work, but you will receive a git warning to update the name of the directory to "datasm".

Per https://stackoverflow.com/a/60082006:

> Renaming the Local Repo Name
>
> As others have mentioned, the local "name" of your repo is typically considered to be the root folder/directory name, and you can change that, move, or copy the folder to any location and it will not affect the repo at all.
> 
> Git is designed to only worry about files inside the root folder.

**Pull request merged**
